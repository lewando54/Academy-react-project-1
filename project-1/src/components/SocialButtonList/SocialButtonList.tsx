import SocialButton from "../SocialButton/SocialButton";

function SocialButtonList(){
    const socials = [
        {id: 0, color: '#e90800', icon: 'google', href: 'https://google.pl/'},
        {id: 1, color: '#4768aa', icon: 'facebook', href: 'https://google.pl/'},
        {id: 2, color: '#0068c1', icon: 'linkedin', href: 'https://google.pl/'}
    ];

    const socialItems = socials.map(social => 
        <SocialButton key={social.id} color={social.color} href={social.href} icon={social.icon}/>
    )

    return (
        <div className="w-full flex justify-center gap-4">
            {socialItems}
        </div>
    )
}

export default SocialButtonList