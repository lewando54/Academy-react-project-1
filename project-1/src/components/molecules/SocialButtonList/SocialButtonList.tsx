import React from 'react'
import SocialButton, { iconProp } from '../../atoms/SocialButton/SocialButton'
import SocialButtonListStyle from './SocialButtonList.style'

export type TSocialsArray = Array<{ id: number, color: string, icon: iconProp, href: URL['href'], testId?: string }>
interface ISocialButtonList {
  socials: TSocialsArray
  testId: string
}

function SocialButtonList ({ socials, testId }: ISocialButtonList): React.ReactElement {
  const socialItems = socials.map(social =>
        <SocialButton key={social.id} color={social.color} href={social.href} icon={social.icon} testId={testId + social.id.toString()}/>
  )

  return (
        <SocialButtonListStyle data-testid={testId}>
            {socialItems}
        </SocialButtonListStyle>
  )
}

export default SocialButtonList
