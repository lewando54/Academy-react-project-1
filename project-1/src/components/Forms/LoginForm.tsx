import React, { useRef} from 'react'
import {Link} from 'react-router-dom'
import Button from "../Button/Button"
import Input from "../Input/Input"
import Or from "../DividerWithText/DividerWithText"
import SocialButtonList from '../SocialButtonList/SocialButtonList'
import Checkbox from '../Checkbox/Checkbox'
import FormContainer from './FormContainer.style'
import FormTitle from './FormTitle.style'

function Form(){
    const emailInputRef = useRef<HTMLInputElement>(null);
    const passwordInputRef = useRef<HTMLInputElement>(null);
    const rememberMeCheckboxRef = useRef<HTMLInputElement>(null);
    // use ref
    // ForwardRef
    // use imperative handle

    const handleLoginSubmit = () => {
        //Make request to backend
        console.log('Login:');
        console.log(emailInputRef.current!.value);
        console.log(passwordInputRef.current!.value);
        console.log(rememberMeCheckboxRef.current!.checked);
        console.log('---------------');
    }

    // przepisać tutaj onInputChange ()=>
    // Dodać ESLint - .eslintrc.js
    // <Input type="password" id="password" name="Password" onInputChange={onInputChange}/>

    return (
        <FormContainer>
            <FormTitle>Login</FormTitle>
            <Input innerRef={emailInputRef} type="email" id="email" name="Email"/>
            <Input innerRef={passwordInputRef} type="password" id="password" name="Password"/>
            <Checkbox innerRef={rememberMeCheckboxRef} id="remember" name="Remember">Remember me?</Checkbox>
            <Button color={"primary"} onClick={handleLoginSubmit}>Login</Button>
            <a className='w-full text-sm text-gray-400 text-right cursor-pointer -mt-2'>Forgot password?</a>
            <Or text='OR'/>
            <SocialButtonList />
            <div className='flex gap-1 text-sm text-gray-500'><p>Need an account?</p><Link to='/register' className='underline'>SIGN UP</Link></div>
        </FormContainer>
    )
}

export default Form