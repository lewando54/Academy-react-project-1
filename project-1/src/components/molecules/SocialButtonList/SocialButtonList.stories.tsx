/* eslint-disable @typescript-eslint/consistent-type-assertions */
import React from 'react'
import { ComponentStory, ComponentMeta } from '@storybook/react'

import SocialButtonList from './SocialButtonList'

// More on default export: https://storybook.js.org/docs/react/writing-stories/introduction#default-export
export default {
  component: SocialButtonList,
  // More on argTypes: https://storybook.js.org/docs/react/api/argtypes
  argTypes: {
    socials: { control: 'array' }
  }

} as ComponentMeta<typeof SocialButtonList>

// More on component templates: https://storybook.js.org/docs/react/writing-stories/introduction#using-args
const Template: ComponentStory<typeof SocialButtonList> = (args) => <SocialButtonList {...args} />

export const Primary = Template.bind({})

Primary.args = {
  socials: [
    { id: 0, color: '#e90800', icon: 'google', href: 'https://google.pl/' },
    { id: 1, color: '#4768aa', icon: 'facebook', href: 'https://google.pl/' },
    { id: 2, color: '#0068c1', icon: 'linkedin', href: 'https://google.pl/' }
  ]
}
// More on args: https://storybook.js.org/docs/react/writing-stories/args
