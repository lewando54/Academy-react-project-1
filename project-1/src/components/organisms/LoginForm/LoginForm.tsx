import React, { useRef, useImperativeHandle, forwardRef } from 'react'
import Button from '../../atoms/Button/Button'
import InputWithLabel from '../../molecules/InputWithLabel/InputWithLabel'
import Divider from '../../atoms/Divider/Divider'
import SocialButtonList, { TSocialsArray } from '../../molecules/SocialButtonList/SocialButtonList'
import CheckboxWithLabel from '../../molecules/CheckboxWithLabel/CheckboxWithLabel'
import FormContainer from './LoginForm.style'
import Heading from '../../atoms/Heading/Heading'
import styled from 'styled-components'
import TextWithLink from '../../atoms/TextWithLink/TextWithLink'

const ForgotPasswordElement = styled.a`
    width: 100%;
    font-size: 0.9rem;
    color: #aaa;
    margin-top: -0.5rem;
    text-align: right;
    cursor: pointer;
`

export interface ILoginProps {
  socials: TSocialsArray
  onSubmit: (e: React.FormEvent<HTMLButtonElement>) => void
  testId: string
}

export interface ILoginRef {
  email: HTMLInputElement | null
  password: HTMLInputElement | null
  rememberMe: HTMLInputElement | null
}

function Form (props: ILoginProps, ref: React.Ref<ILoginRef>): React.ReactElement {
  const emailInputRef = useRef<HTMLInputElement>(null)
  const passwordInputRef = useRef<HTMLInputElement>(null)
  const rememberMeCheckboxRef = useRef<HTMLInputElement>(null)

  // RozbiÄ‡ focus na dwie metody
  useImperativeHandle(ref, () => {
    return {
      email: emailInputRef.current,
      password: passwordInputRef.current,
      rememberMe: rememberMeCheckboxRef.current,
      focus () {
        if (emailInputRef.current != null) {
          emailInputRef.current.focus()
        }
      },
      clear () {
        if (emailInputRef.current != null &&
          passwordInputRef.current != null &&
          rememberMeCheckboxRef.current != null) {
          emailInputRef.current.value = ''
          passwordInputRef.current.value = ''
          rememberMeCheckboxRef.current.checked = false
        }
      }
    }
  }, [])

  // TestId na sztywno

  return (
        <FormContainer>
            <Heading testId={props.testId + '_heading'}>Login</Heading>
            <InputWithLabel ref={emailInputRef} type="email" id="email" name="Email" required testId={props.testId + '_email'}>Email</InputWithLabel>
            <InputWithLabel ref={passwordInputRef} type="password" id="password" name="Password" required testId={props.testId + '_password'}>Password</InputWithLabel>
            <CheckboxWithLabel ref={rememberMeCheckboxRef} id="remember" name="Remember" testId={props.testId + '_remember'} color='red'>Remember me?</CheckboxWithLabel>
            <Button color={'primary'} onClick={props.onSubmit} testId={props.testId + '_submit'}>Login</Button>
            <ForgotPasswordElement>Forgot password?</ForgotPasswordElement>
            <Divider text='OR' testId={props.testId + '_divider'}/>
            <SocialButtonList socials={props.socials} testId={props.testId + '_socialBtnList'}/>
            <TextWithLink paragraphText={'Need an account?'} href={'/register'} anchorText={'SIGN UP'} testId={props.testId + '_textWithLink'} anchorTagTestId={props.testId + '_anchorTextWithLink'}/>
        </FormContainer>
  )
}

export default forwardRef(Form)
