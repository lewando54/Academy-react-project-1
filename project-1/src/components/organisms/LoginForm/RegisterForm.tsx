import React, { useRef } from 'react'
import Button from '../../atoms/Button/Button'
import InputWithLabel from '../../molecules/InputWithLabel/InputWithLabel'
import Or from '../../atoms/DividerWithText/DividerWithText'
import SocialButtonList from '../../molecules/SocialButtonList/SocialButtonList'
import FormContainer from './FormContainer.style'
import FormTitle from '../../atoms/TitleHeading/TitleHeading.style'
import TextWithLink from '../../atoms/TextWithLink/TextWithLink'

function RegisterForm (): React.ReactElement {
  const emailInputRef = useRef<HTMLInputElement>(null)
  const passwordInputRef = useRef<HTMLInputElement>(null)

  const handleRegisterSubmit = (e: React.FormEvent<HTMLButtonElement>): void => {
    // Make request to backend
    console.log('Register:')
    console.log(emailInputRef.current?.value)
    console.log(passwordInputRef.current?.value)
    console.log('---------------')
  }

  return (
        <FormContainer>
            <FormTitle>Sign up</FormTitle>
            <InputWithLabel innerRef={emailInputRef} type="email" id="email" name="Email" required>Email</InputWithLabel>
            <InputWithLabel innerRef={passwordInputRef} type="password" id="password" name="Password" required>Password</InputWithLabel>
            <Button color="primary" onClick={handleRegisterSubmit}>Sign up</Button>
            <Or text='OR'/>
            <SocialButtonList />
            <TextWithLink pText={'Already a user?'} href={'/login'} anchorText={'SIGN UP'}/>
        </FormContainer>
  )
}

export default RegisterForm
