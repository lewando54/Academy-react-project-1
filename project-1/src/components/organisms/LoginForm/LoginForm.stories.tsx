/* eslint-disable @typescript-eslint/consistent-type-assertions */
import React from 'react'
import { ComponentStory, ComponentMeta } from '@storybook/react'

import { withRouter } from 'storybook-addon-react-router-v6';
import LoginForm from './LoginForm'

// More on default export: https://storybook.js.org/docs/react/writing-stories/introduction#default-export
export default {
  component: LoginForm,
  // More on argTypes: https://storybook.js.org/docs/react/api/argtypes
  argTypes: {
    socials: { control: 'array' },
    onSubmit: { action: 'submitted form' }
  },
  decorators: [withRouter]
} as ComponentMeta<typeof LoginForm>

// More on component templates: https://storybook.js.org/docs/react/writing-stories/introduction#using-args
const Template: ComponentStory<typeof LoginForm> = (args) => <LoginForm {...args} />

export const Primary = Template.bind({})

Primary.args = {
  socials: [
    { id: 0, color: '#e90800', icon: 'google', href: 'https://google.pl/' },
    { id: 1, color: '#4768aa', icon: 'facebook', href: 'https://google.pl/' },
    { id: 2, color: '#0068c1', icon: 'linkedin', href: 'https://google.pl/' }
  ]
}
// More on args: https://storybook.js.org/docs/react/writing-stories/args
