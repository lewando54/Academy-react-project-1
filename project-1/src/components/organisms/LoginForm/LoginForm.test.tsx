import React, { createRef } from 'react'
import { render, fireEvent } from '@testing-library/react'
import '@testing-library/jest-dom'
import { MemoryRouter } from 'react-router-dom'
import LoginForm from './LoginForm'
import { TSocialsArray } from '../../molecules/SocialButtonList/SocialButtonList'

describe('LoginForm', () => {
  interface ILoginFormRef {
    email: HTMLInputElement
    password: HTMLInputElement
    rememberMe: HTMLInputElement
    focus: () => void
    clear: () => void
  }

  const socials: TSocialsArray = [
    { id: 1, color: 'blue', href: 'Facebook', icon: 'facebook' },
    { id: 2, color: 'red', href: 'Google', icon: 'google' }
  ]
  const onSubmit = jest.fn()
  const testId = 'login-form'

  it('should render all form elements', () => {
    const { getByLabelText, getByText, getByRole } = render(
      <MemoryRouter><LoginForm socials={socials} onSubmit={onSubmit} testId={testId} /></MemoryRouter>
    )

    expect(getByLabelText('Email')).toBeInTheDocument()
    expect(getByLabelText('Password')).toBeInTheDocument()
    expect(getByText('Remember me?')).toBeInTheDocument()
    expect(getByRole('button')).toBeInTheDocument()
    expect(getByText('Forgot password?')).toBeInTheDocument()
    expect(getByText('OR')).toBeInTheDocument()
    expect(getByText('Need an account?')).toBeInTheDocument()
    expect(getByText('SIGN UP')).toBeInTheDocument()
  })

  it('should call onSubmit when the form is submitted', () => {
    const { getByTestId } = render(
      <MemoryRouter><LoginForm socials={socials} onSubmit={onSubmit} testId={testId} /></MemoryRouter>
    )
    const submitButton = getByTestId(`${testId}_submit`)
    fireEvent.click(submitButton)
    expect(onSubmit).toHaveBeenCalled()
  })

  it('should clear all inputs when the clear method is called', () => {
    const formRef = createRef<ILoginFormRef>()

    const { getByLabelText } = render(
      <MemoryRouter><LoginForm socials={socials} onSubmit={onSubmit} ref={formRef} testId={testId} /></MemoryRouter>
    )

    const emailInput = getByLabelText('Email')
    const passwordInput = getByLabelText('Password')
    const rememberMeCheckbox = getByLabelText('Remember me?')
    fireEvent.change(emailInput, { target: { value: 'test@example.com' } })
    fireEvent.change(passwordInput, { target: { value: 'password' } })
    fireEvent.click(rememberMeCheckbox)

    formRef.current?.clear()

    expect(formRef.current?.email.value).toBe('')
    expect(formRef.current?.password.value).toBe('')
    expect(formRef.current?.rememberMe.checked).toBe(false)
  })

  it('should focus on the email input when the focus method is called', () => {
    const formRef = createRef<ILoginFormRef>()

    const { getByLabelText } = render(
      <MemoryRouter><LoginForm socials={socials} onSubmit={onSubmit} ref={formRef} testId={testId} /></MemoryRouter>
    )

    const emailInput = getByLabelText('Email')
    const passwordInput = getByLabelText('Password')
    fireEvent.focus(passwordInput)

    formRef.current?.focus()

    expect(document.activeElement).toBe(emailInput)
  })
})
