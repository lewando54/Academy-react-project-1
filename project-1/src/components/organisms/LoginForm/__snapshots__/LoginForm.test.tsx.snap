// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LoginForm should render correctly 1`] = `
<div
  className="sc-jrcTuL fdFQHl"
>
  <h1
    className="sc-kDvujY gJTozq"
    data-testid="test_heading"
  >
    Login
  </h1>
  <div
    className="sc-dkrFOg fkwlLo"
  >
    <label
      data-testid="test_email"
      htmlFor="email"
    >
      Email
    </label>
    <input
      className="sc-gswNZR ftHZNR"
      data-testid="test_email"
      id="email"
      name="Email"
      required={true}
      type="email"
    />
  </div>
  <div
    className="sc-dkrFOg fkwlLo"
  >
    <label
      data-testid="test_password"
      htmlFor="password"
    >
      Password
    </label>
    <input
      className="sc-gswNZR ftHZNR"
      data-testid="test_password"
      id="password"
      name="Password"
      required={true}
      type="password"
    />
  </div>
  <div
    className="sc-pyfCe dRbsPv"
  >
    <label
      data-testid="test_remember"
      htmlFor="remember"
    >
      Remember me?
    </label>
    <input
      className="sc-ftTHYK jigCOa"
      data-testid="test_remember"
      id="remember"
      type="checkbox"
    />
  </div>
  <button
    className="sc-bcXHqe iJjCPT"
    color="primary"
    data-testid="test_submit"
    onClick={[MockFunction]}
  >
    Login
  </button>
  <a
    className="sc-eDWCr jOPjxA"
  >
    Forgot password?
  </a>
  <div
    className="sc-eDvSVe llSYDy"
    data-testid="test_divider"
  >
    <hr
      className="sc-hLBbgP fdaVyi"
    />
    <p
      className="sc-jSUZER fxgfMS"
    >
      OR
    </p>
    <hr
      className="sc-hLBbgP fdaVyi"
    />
  </div>
  <div
    className="sc-iBYQkv qXJyF"
    data-testid="test_socialBtnList"
  >
    <a
      className="sc-gKPRtg fXEYBM"
      color="#e90800"
      data-testid="test_socialBtnList0"
      href="https://google.pl/"
    >
      <svg
        fill="currentColor"
        height="1em"
        stroke="currentColor"
        strokeWidth="0"
        style={
          {
            "color": undefined,
          }
        }
        viewBox="0 0 488 512"
        width="1em"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M488 261.8C488 403.3 391.1 504 248 504 110.8 504 0 393.2 0 256S110.8 8 248 8c66.8 0 123 24.5 166.3 64.9l-67.5 64.9C258.5 52.6 94.3 116.6 94.3 256c0 86.5 69.1 156.6 153.7 156.6 98.2 0 135-70.4 140.8-106.9H248v-85.3h236.1c2.3 12.7 3.9 24.9 3.9 41.4z"
        />
      </svg>
    </a>
    <a
      className="sc-gKPRtg ielPGY"
      color="#4768aa"
      data-testid="test_socialBtnList1"
      href="https://google.pl/"
    >
      <svg
        fill="currentColor"
        height="1em"
        stroke="currentColor"
        strokeWidth="0"
        style={
          {
            "color": undefined,
          }
        }
        viewBox="0 0 320 512"
        width="1em"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M279.14 288l14.22-92.66h-88.91v-60.13c0-25.35 12.42-50.06 52.24-50.06h40.42V6.26S260.43 0 225.36 0c-73.22 0-121.08 44.38-121.08 124.72v70.62H22.89V288h81.39v224h100.17V288z"
        />
      </svg>
    </a>
    <a
      className="sc-gKPRtg cNQgFw"
      color="#0068c1"
      data-testid="test_socialBtnList2"
      href="https://google.pl/"
    >
      <svg
        fill="currentColor"
        height="1em"
        stroke="currentColor"
        strokeWidth="0"
        style={
          {
            "color": undefined,
          }
        }
        viewBox="0 0 448 512"
        width="1em"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M100.28 448H7.4V148.9h92.88zM53.79 108.1C24.09 108.1 0 83.5 0 53.8a53.79 53.79 0 0 1 107.58 0c0 29.7-24.1 54.3-53.79 54.3zM447.9 448h-92.68V302.4c0-34.7-.7-79.2-48.29-79.2-48.29 0-55.69 37.7-55.69 76.7V448h-92.78V148.9h89.08v40.8h1.3c12.4-23.5 42.69-48.3 87.88-48.3 94 0 111.28 61.9 111.28 142.3V448z"
        />
      </svg>
    </a>
  </div>
  <div
    className="sc-ipEyDJ aZMVu"
    data-testid="test_textWithLink"
  >
    <p>
      Need an account?
    </p>
    <a
      className="sc-csuSiG NkRvB"
      data-testid="test_anchorTextWithLink"
      onClick={[Function]}
    >
      SIGN UP
    </a>
  </div>
</div>
`;
